### PR TITLE
LF-122: Refactor balance service

### PR DESCRIPTION
**Pull request by mathiasmoeller**
[LF-122](https://lifi.atlassian.net/jira/software/projects/LF/boards/1?selectedIssue=LF-122)

# What has been done
* Split up fetching balances from chains and from third party services
* Move chain-based balances checks into LiFi SDK
* Refactor usages of old balance service

# What needs to be done
* Debugging like crazy...
  * fetching data from debank seems to work
  * [x] `getBalancesFromProviderUsingMulticall` in `services/LIFI/balances/utils` has issues. The multicall throws an error which is caught by the watcher. Couldn't figure out yet what goes wrong. The implementation didn't change much here
  * [x] `Dashboard.tsx` loads the data from Debank. Then it saves it in a local portfolio. Here we have a missmatch because we use `ChainId` and sometimes `ChainKey` to identify chains. Need to go through the code and switch to one usage
  * [x] Test swapping on Lifi 
  * [x] Test swapping on XPollinate
  * [x] Decide on a way on how to unit/integration test the new balance service in the SDK

### CODE REVIEW

File: src/components/Swapping.tsx [hunk 1]
```diff
@@ -17,10 +17,10 @@ import paraswapIcon from '../assets/icons/paraswap.png'
 import walletIcon from '../assets/wallet.png'
 import { oneInch } from '../services/1Inch'
 import { AnySwapExecutionManager } from '../services/anyswap.execute'
-import { getBalancesForWallet } from '../services/balanceService'
 import { CbridgeExecutionManager } from '../services/cbridge.execute'
 import { HopExecutionManager } from '../services/hop.execute'
 import { HorizonExecutionManager } from '../services/horizon.execute'
+import Lifi from '../services/LIFI/Lifi'
```
Comment (Addminus) [inline src/components/Swapping.tsx:23]: The SDK does not have the config file the normal balance service has. I understand, that it is not necessary for the SDK to allow for maximum freedom.
However, blocking malicious tokens on the user-facing site seems like a nice feature.

**Review by Addminus — COMMENTED** [review]
Any specific reasons, why the balance service in LIFI does not use Covalent or Debank?

Also `Swapping.tsx` is using the balance service from the SDK and thereby loses the blacklist for scam tokens. I think this feature is quite nice for the user-facing website. Would it be possible to integrate the SDK balance-service functions with the 'normal' balance service function, creating a flow similar to this:

Swapping.txs  imports balance-service imports SDK-balance-service

The normal balance service would then act as a gatekeeper and would NOT forward requests for malicious tokens to the SDK-balance-service

That would hide those tokens from the user and might reduce accidental transfers.

**Comment by maxklenk** [comment]
Querying balances using aggregators apis like DeBank/Covalent and accessing them on the chain e.g. using multicall have different advantages/disadvantages.
```
Aggregators:
+ get all token the user have
+ one request only
- my contain fake tokens
- balances may be outdated
- balances are rounded (DeBank)
- not available for all chains

Chain:
+ up to date balances
- we have to request specific tokens
```

We started with Aggregators but figured out that the disadvantages create problems. e.g. we think the user has more token than he actually has because the amounts are outdated/rounded. Which results in failing transactions. With Multicall we are able to query multiple tokens in one request as well, since we request a specific list of tokens we don't need any blacklisting - we use whitelisting using token-lists.

For the Dashboard on the other hand we want to display as many token as possible, so the aggregators are great. 

In the future we may add a combination of both approaches: 
- initially load all tokens a user has using aggregators
- filter out using blacklist
- query real balances using multicall

File: src/components/nxtp/SwapFormNxtp.tsx [hunk 2]
```diff
@@ -273,9 +273,9 @@ const SwapFormNxtp = ({
               <div>
                 {depositToken &&
                   balances &&
-                  (balances[depositChain] ?? [])
-                    .find((p) => p.id === depositToken)
-                    ?.amount.toFixed(4)}
+                  new BigNumber(
+                    (balances[depositChain] ?? []).find((p) => p.id === depositToken)?.amount || 0,
```
[the remark this answers is not in the document]
Comment (maxklenk) [reply src/components/nxtp/SwapFormNxtp.tsx:277]: yes that is no problem

**Comment by Addminus** [comment]
> Querying balances using aggregators apis like DeBank/Covalent and accessing them on the chain e.g. using multicall have different advantages/disadvantages.
> 
> ```
> Aggregators:
> + get all token the user have
> + one request only
> - my contain fake tokens
> - balances may be outdated
> - balances are rounded (DeBank)
> - not available for all chains
> 
> Chain:
> + up to date balances
> - we have to request specific tokens
> ```
> 
> We started with Aggregators but figured out that the disadvantages create problems. e.g. we think the user has more token than he actually has because the amounts are outdated/rounded. Which results in failing transactions. With Multicall we are able to query multiple tokens in one request as well, since we request a specific list of tokens we don't need any blacklisting - we use whitelisting using token-lists.
> 
> For the Dashboard on the other hand we want to display as many token as possible, so the aggregators are great.
> 
> In the future we may add a combination of both approaches:
> 
> * initially load all tokens a user has using aggregators
> * filter out using blacklist
> * query real balances using multicall

Alright! Thanks for clarifying!

**Pull request merged**